In [1]:
import pandas as pd

In [2]:
df_tech_500=pd.read_csv("../processed/tech_sp500_stable_extended_tickers.csv")

Here, we are using stable companies between 2014 to 2024.

In [4]:
df_tech_500[df_tech_500["stable 14 to 24"]==True]

Unnamed: 0      cik                                  name entityType  \
0            0   320193                            Apple Inc.  operating   
1            1   789019                        MICROSOFT CORP  operating   
2            2  1045810                           NVIDIA CORP  operating   
3            3  1652044                         Alphabet Inc.  operating   
4            4  1326801                  Meta Platforms, Inc.  operating   
5            5  1730168                         Broadcom Inc.  operating   
6            6  1341439                           ORACLE CORP  operating   
7            7  1108524                      Salesforce, Inc.  operating   
8            8   796343                            ADOBE INC.  operating   
10          10    97745         THERMO FISHER SCIENTIFIC INC.  operating   
12          12   858877                   CISCO SYSTEMS, INC.  operating   
13          13   313616                     DANAHER CORP /DE/  operating   
14          14   896878                           INTUIT INC.  operating   
15          15    97476                 TEXAS INSTRUMENTS INC  operating   
16          16    51143  INTERNATIONAL BUSINESS MACHINES CORP  operating   
17          17     6951             APPLIED MATERIALS INC /DE  operating   
19          19   732712            VERIZON COMMUNICATIONS INC  operating   
20          20    18230                       CATERPILLAR INC  operating   
21          21  1166691                          COMCAST CORP  operating   
22          22   732717                             AT&T INC.  operating   
23          23   723125                 MICRON TECHNOLOGY INC  operating   
24          24  1551182                        Eaton Corp plc  operating   
25          25   707549                     LAM RESEARCH CORP  operating   
27          27     6281                    ANALOG DEVICES INC  operating   
29          29  1613103                         Medtronic plc  operating   
30          30     8670         AUTOMATIC DATA PROCESSING INC  operating   
31          31   315189                            DEERE & CO  operating   
34          34   820313                    AMPHENOL CORP /DE/  operating   
39          39    49826               ILLINOIS TOOL WORKS INC  operating   
43          43   882835                ROPER TECHNOLOGIES INC  operating   
45          45   769397                        Autodesk, Inc.  operating   
47          47   833444    Johnson Controls International plc  operating   
49          49   827054              MICROCHIP TECHNOLOGY INC  operating   
50          50    26172                           CUMMINS INC  operating   
51          51   712515                  ELECTRONIC ARTS INC.  operating   
53          53  1037868                           AMETEK INC/  operating   
54          54  1058290   COGNIZANT TECHNOLOGY SOLUTIONS CORP  operating   
57          57    47217                                HP INC  operating   
60          60  1524472                            Xylem Inc.  operating   
61          61  1024478              ROCKWELL AUTOMATION, INC  operating   
63          63  1002047                          NetApp, Inc.  operating   
67          67    29905                            DOVER Corp  operating   
74          74  1137789       Seagate Technology Holdings plc  operating   
75          75   106040                  WESTERN DIGITAL CORP  operating   
81          81  1754301                              Fox Corp  operating   
82          82  1014473                       VERISIGN INC/CA  operating   
88          88    77360                           PENTAIR plc  operating   
91          91  1043604                  JUNIPER NETWORKS INC  operating   
95          95  1048695                              F5, INC.  operating   
98          98    46080                          HASBRO, INC.  operating   

       sic                                     sicDescription  \
0   3571.0                               Electronic Computers 

In [12]:
lst_naics_str=list(df_tech_500.naics)

In [13]:
lst_naics_set=[set(ele_str.split(",")) for ele_str in lst_naics_str]

below is a function that union a list of sets.

In [15]:
def lst_set_union(lst_set): 
    union_set=set()
    for ele_set in lst_set: 
        union_set=union_set.union(ele_set)
    return union_set

In [16]:
set_naics=lst_set_union(lst_naics_set)

Below is the naics we care about, it seems a bit too long, and I only have a normal machine, so lets shorten it by industry. 

In [17]:
set_naics

{'314999',
 '332216',
 '332323',
 '332410',
 '332710',
 '332813',
 '332999',
 '333111',
 '333120',
 '333131',
 '333132',
 '333242',
 '333249',
 '333316',
 '333318',
 '333414',
 '333415',
 '333611',
 '333618',
 '333914',
 '333922',
 '333923',
 '333992',
 '333999',
 '334111',
 '334112',
 '334118',
 '334210',
 '334220',
 '334310',
 '334413',
 '334418',
 '334419',
 '334510',
 '334512',
 '334513',
 '334514',
 '334515',
 '334517',
 '334519',
 '334613',
 '334614',
 '335121',
 '335122',
 '335129',
 '335999',
 '336390',
 '336510',
 '336991',
 '339112',
 '339930',
 '339940',
 '511210',
 '515120',
 '515210',
 '517311',
 '517312',
 '517911',
 '518210',
 '541511',
 '541512'}

In [21]:
set_naics_short=set()

for ele_str in set_naics: 
    set_naics_short.add(ele_str[:3]+"---")

In [22]:
set_naics_short

{'314---',
 '332---',
 '333---',
 '334---',
 '335---',
 '336---',
 '339---',
 '511---',
 '515---',
 '517---',
 '518---',
 '541---'}

Above is a list of shorten naics we care about, it might not be as detailed, but at least reasonable amount 

In [19]:
df_naics_ppi=pd.read_table("../raw/ppi_and_cpi/ppi_ind_series.txt")

In [20]:
df_naics_ppi

series_id                      industry_code product_code seasonal  \
0     PCU1133--1133--                       1133--       1133--        U   
1     PCU11331-11331-                       11331-       11331-        U   
2     PCU113310113310                       113310       113310        U   
3     PCU113310113310M                      113310      113310M        U   
4     PCU113310113310MM                     113310     113310MM        U   
...                              ...           ...          ...      ...   
4733  PCUATRNWRATRNWR                       ATRNWR       ATRNWR        U   
4734  PCUATTDSVATTDSV                       ATTDSV       ATTDSV        U   
4735  PCUAWHLTRAWHLTR                       AWHLTR       AWHLTR        U   
4736  PCUOMFG--OMFG--                       OMFG--       OMFG--        U   
4737  PCUOMIN--OMIN--                       OMIN--       OMIN--        U   

      base_date                                       series_title  \
0        198112  PPI industry group data for Logging, not seaso...   
1        198112  PPI industry data for Logging, not seasonally ...   
2        198112  PPI industry data for Logging, not seasonally ...   
3        198606  PPI industry data for Logging-Miscellaneous re...   
4        198606  PPI industry data for Logging-Miscellaneous re...   
...         ...                                                ...   
4733     200612  PPI industry data for Transportation and wareh...   
4734     200612  PPI industry data for Services less trade, tra...   
4735     200612  PPI industry data for Total wholesale trade in...   
4736     198412  PPI industry group data for Total manufacturin...   
4737     198412  PPI industry group data for Total mining indus...   

      footnote_codes  begin_year begin_period  end_year end_period  
0                NaN        1981          M12      2025        M01  
1                NaN        1981          M12      2025        M01  
2                NaN        1981          M12      2025        M01  
3                NaN        1986          M06      2022        M07  
4                NaN        1986          M06      1998        M08  
...              ...         ...          ...       ...        ...  
4733             NaN        2006          M12      2025        M01  
4734             NaN        2006          M12      2025        M01  
4735             NaN        2006          M12      2025        M01  
4736             NaN        1984          M12      2025        M01  
4737             NaN        1984          M12      2025        M01  

[4738 rows x 11 columns]

In [26]:
df_naics_ppi[df_naics_ppi["industry_code"].isin(set_naics_short)]

series_id                      industry_code product_code seasonal  \
819   PCU314---314---                       314---       314---        U   
2071  PCU332---332---                       332---       332---        U   
2382  PCU333---333---                       333---       333---        U   
2753  PCU334---334---                       334---       334---        U   
2912  PCU335---335---                       335---       335---        U   
3054  PCU336---336---                       336---       336---        U   
3375  PCU339---339---                       339---       339---        U   
4183  PCU517---517---                       517---       517---        U   

      base_date                                       series_title  \
819      200312  PPI industry sub-sector data for Textile produ...   
2071     198412  PPI industry sub-sector data for Fabricated me...   
2382     200312  PPI industry sub-sector data for Machinery man...   
2753     200312  PPI industry sub-sector data for Computer & el...   
2912     200312  PPI industry sub-sector data for Electrical eq...   
3054     200312  PPI industry sub-sector data for Transportatio...   
3375     200312  PPI industry sub-sector data for Miscellaneous...   
4183     200312  PPI industry sub-sector data for Telecommunica...   

      footnote_codes  begin_year begin_period  end_year end_period  
819              NaN        2003          M12      2025        M01  
2071             NaN        1984          M12      2025        M01  
2382             NaN        2003          M12      2025        M01  
2753             NaN        2003          M12      2025        M01  
2912             NaN        2003          M12      2025        M01  
3054             NaN        2003          M12      2025        M01  
3375             NaN        2003          M12      2025        M01  
4183             NaN        2003          M12      2025        M01

There are some missing ones: 541---, 518---, 515---, 511---, we need to include one more digit in those naics to try again. 

In [29]:
set_naics_short=set_naics_short.union({"5415--","5182--","5151--","5152--","5112--"})

In [30]:
df_naics_ppi[df_naics_ppi["industry_code"].isin(set_naics_short)]

series_id                      industry_code product_code seasonal  \
819   PCU314---314---                       314---       314---        U   
2071  PCU332---332---                       332---       332---        U   
2382  PCU333---333---                       333---       333---        U   
2753  PCU334---334---                       334---       334---        U   
2912  PCU335---335---                       335---       335---        U   
3054  PCU336---336---                       336---       336---        U   
3375  PCU339---339---                       339---       339---        U   
4183  PCU517---517---                       517---       517---        U   
4209  PCU5182--5182--                       5182--       5182--        U   

      base_date                                       series_title  \
819      200312  PPI industry sub-sector data for Textile produ...   
2071     198412  PPI industry sub-sector data for Fabricated me...   
2382     200312  PPI industry sub-sector data for Machinery man...   
2753     200312  PPI industry sub-sector data for Computer & el...   
2912     200312  PPI industry sub-sector data for Electrical eq...   
3054     200312  PPI industry sub-sector data for Transportatio...   
3375     200312  PPI industry sub-sector data for Miscellaneous...   
4183     200312  PPI industry sub-sector data for Telecommunica...   
4209     200312  PPI industry group data for Data processing an...   

      footnote_codes  begin_year begin_period  end_year end_period  
819              NaN        2003          M12      2025        M01  
2071             NaN        1984          M12      2025        M01  
2382             NaN        2003          M12      2025        M01  
2753             NaN        2003          M12      2025        M01  
2912             NaN        2003          M12      2025        M01  
3054             NaN        2003          M12      2025        M01  
3375             NaN        2003          M12      2025        M01  
4183             NaN        2003          M12      2025        M01  
4209             NaN        2003          M12      2025        M01

still missing: 5415--, 5151--, 5152--, 5112--.

5415--: Computer system design services (could not find similar, this is very annoying)

515---: broad casting (similar to 516---)

5112--: software publisher (similar to 5132--)

In [31]:
set_naics_short=set_naics_short.union({"516---","5132--"})

In [33]:
df_ppi_care=df_naics_ppi[df_naics_ppi["industry_code"].isin(set_naics_short)]

In [52]:
df_ppi_care

series_id                      industry_code product_code seasonal  \
819   PCU314---314---                       314---       314---        U   
2071  PCU332---332---                       332---       332---        U   
2382  PCU333---333---                       333---       333---        U   
2753  PCU334---334---                       334---       334---        U   
2912  PCU335---335---                       335---       335---        U   
3054  PCU336---336---                       336---       336---        U   
3375  PCU339---339---                       339---       339---        U   
4145  PCU5132--5132--                       5132--       5132--        U   
4158  PCU516---516---                       516---       516---        U   
4183  PCU517---517---                       517---       517---        U   
4209  PCU5182--5182--                       5182--       5182--        U   

      base_date                                       series_title  \
819      200312  PPI industry sub-sector data for Textile produ...   
2071     198412  PPI industry sub-sector data for Fabricated me...   
2382     200312  PPI industry sub-sector data for Machinery man...   
2753     200312  PPI industry sub-sector data for Computer & el...   
2912     200312  PPI industry sub-sector data for Electrical eq...   
3054     200312  PPI industry sub-sector data for Transportatio...   
3375     200312  PPI industry sub-sector data for Miscellaneous...   
4145     200312  PPI industry group data for Software publisher...   
4158     200312  PPI industry sub-sector data for Broadcasting,...   
4183     200312  PPI industry sub-sector data for Telecommunica...   
4209     200312  PPI industry group data for Data processing an...   

      footnote_codes  begin_year begin_period  end_year end_period  
819              NaN        2003          M12      2025        M01  
2071             NaN        1984          M12      2025        M01  
2382             NaN        2003          M12      2025        M01  
2753             NaN        2003          M12      2025        M01  
2912             NaN        2003          M12      2025        M01  
3054             NaN        2003          M12      2025        M01  
3375             NaN        2003          M12      2025        M01  
4145             NaN        2003          M12      2025        M01  
4158             NaN        2003          M12      2025        M01  
4183             NaN        2003          M12      2025        M01  
4209             NaN        2003          M12      2025        M01

Now we try to download the data

In [53]:
df_ppi_care.columns

Index(['series_id                     ', 'industry_code', 'product_code',
       'seasonal', 'base_date', 'series_title', 'footnote_codes', 'begin_year',
       'begin_period', 'end_year', 'end_period'],
      dtype='object')

In [54]:
series_care=list(df_ppi_care["series_id                     "])

In [55]:
series_care=[ele_str[:15] for ele_str in series_care]

In [56]:
series_care

['PCU314---314---',
 'PCU332---332---',
 'PCU333---333---',
 'PCU334---334---',
 'PCU335---335---',
 'PCU336---336---',
 'PCU339---339---',
 'PCU5132--5132--',
 'PCU516---516---',
 'PCU517---517---',
 'PCU5182--5182--']

In [43]:
import requests
import json
import prettytable

In [62]:
headers = {'Content-type': 'application/json'}
data = json.dumps({"seriesid": series_care,"startyear":"2003", "endyear":"2012"})
p = requests.post('https://api.bls.gov/publicAPI/v2/timeseries/data/', data=data, headers=headers)
json_data = json.loads(p.text)
for series in json_data['Results']['series']:
    x=prettytable.PrettyTable(["series id","year","period","value","footnotes"])
    seriesId = series['seriesID']
    for item in series['data']:
        year = item['year']
        period = item['period']
        value = item['value']
        footnotes=""
        for footnote in item['footnotes']:
            if footnote:
                footnotes = footnotes + footnote['text'] + ','
        if 'M01' <= period <= 'M12':
            x.add_row([seriesId,year,period,value,footnotes[0:-1]])
    output = open("../raw/ppi_and_cpi/"+seriesId + '(03-12).txt','w')
    output.write (x.get_string())
    output.close()

In [63]:
headers = {'Content-type': 'application/json'}
data = json.dumps({"seriesid": series_care,"startyear":"2012", "endyear":"2020"})
p = requests.post('https://api.bls.gov/publicAPI/v2/timeseries/data/', data=data, headers=headers)
json_data = json.loads(p.text)
for series in json_data['Results']['series']:
    x=prettytable.PrettyTable(["series id","year","period","value","footnotes"])
    seriesId = series['seriesID']
    for item in series['data']:
        year = item['year']
        period = item['period']
        value = item['value']
        footnotes=""
        for footnote in item['footnotes']:
            if footnote:
                footnotes = footnotes + footnote['text'] + ','
        if 'M01' <= period <= 'M12':
            x.add_row([seriesId,year,period,value,footnotes[0:-1]])
    output = open("../raw/ppi_and_cpi/"+seriesId + '(12-20).txt','w')
    output.write (x.get_string())
    output.close()

In [64]:
headers = {'Content-type': 'application/json'}
data = json.dumps({"seriesid": series_care,"startyear":"2020", "endyear":"2025"})
p = requests.post('https://api.bls.gov/publicAPI/v2/timeseries/data/', data=data, headers=headers)
json_data = json.loads(p.text)
for series in json_data['Results']['series']:
    x=prettytable.PrettyTable(["series id","year","period","value","footnotes"])
    seriesId = series['seriesID']
    for item in series['data']:
        year = item['year']
        period = item['period']
        value = item['value']
        footnotes=""
        for footnote in item['footnotes']:
            if footnote:
                footnotes = footnotes + footnote['text'] + ','
        if 'M01' <= period <= 'M12':
            x.add_row([seriesId,year,period,value,footnotes[0:-1]])
    output = open("../raw/ppi_and_cpi/"+seriesId + '(20-25).txt','w')
    output.write (x.get_string())
    output.close()